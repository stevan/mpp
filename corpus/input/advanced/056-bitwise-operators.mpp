# Bitwise AND
my $and = 5 & 3;
my $and2 = 0xFF & 0x0F;

# Bitwise OR
my $or = 5 | 3;
my $or2 = 0x10 | 0x01;

# Bitwise XOR
my $xor = 5 ^ 3;
my $xor2 = 0xFF ^ 0xAA;

# Left shift
my $lshift = 2 << 3;
my $lshift2 = 0x01 << 8;

# Right shift
my $rshift = 16 >> 2;
my $rshift2 = 0xFF00 >> 8;

# Complex bitwise expressions (without NOT operator)
my $complex = ($a & 0xFF) | ($b << 8);
my $combined = ($value & $mask) | $flags;