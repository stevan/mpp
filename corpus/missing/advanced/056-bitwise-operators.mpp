# This file only contains the bitwise NOT operator tests
# All other bitwise operators are supported and in corpus/input/advanced/056-bitwise-operators.mpp

# Bitwise NOT (unsupported)
my $not = ~5;
my $not2 = ~0xFF;

# Complex expressions with bitwise NOT
my $mask = ~(1 << $n);
my $flags = $status & ~$clear_bits | $set_bits;