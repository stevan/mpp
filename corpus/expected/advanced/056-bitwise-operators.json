[
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$and"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "&",
      "left": {
        "type": "Number",
        "value": "5"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$and2"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$or"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "|",
      "left": {
        "type": "Number",
        "value": "5"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$or2"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$xor"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "^",
      "left": {
        "type": "Number",
        "value": "5"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$xor2"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$lshift"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<<",
      "left": {
        "type": "Number",
        "value": "2"
      },
      "right": {
        "type": "Number",
        "value": "3"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$lshift2"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$rshift"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ">>",
      "left": {
        "type": "Number",
        "value": "16"
      },
      "right": {
        "type": "Number",
        "value": "2"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$rshift2"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$complex"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "|",
      "left": {
        "type": "BinaryOp",
        "operator": "&",
        "left": {
          "type": "Variable",
          "name": "$a"
        },
        "right": {
          "type": "Number",
          "value": "0"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": "<<",
        "left": {
          "type": "Variable",
          "name": "$b"
        },
        "right": {
          "type": "Number",
          "value": "8"
        }
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$combined"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "|",
      "left": {
        "type": "BinaryOp",
        "operator": "&",
        "left": {
          "type": "Variable",
          "name": "$value"
        },
        "right": {
          "type": "Variable",
          "name": "$mask"
        }
      },
      "right": {
        "type": "Variable",
        "name": "$flags"
      }
    }
  }
]